### PR TITLE
Implement the framework for entities.

### PR DESCRIPTION
**Pull request by benflexcompute**
Entities implementation completed. The rest of relevant work is how and when `_expand_entities` function is called

### CODE REVIEW

File: flow360/component/simulation/framework/entity_base.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+            entity types.
+
+    """
+
+    stored_entities: List = pd.Field()
+
+    @pd.model_validator(mode="before")
+    @classmethod
+    def format_input_to_list(cls, input: List) -> dict:
+        return dict(stored_entities=input)
```
Comment (maciej-flexcompute) [inline flow360/component/simulation/framework/entity_base.py:104]: here we don't have any validation on `input`, I think. How do we know it is a list? Also, when user provides single string (without []) we can parse it as list for user convenience

Comment (benflexcompute) [reply flow360/component/simulation/framework/entity_base.py]: Added validation

File: tests/simulation_framework/test_entities_v2.py [hunk 2]
```diff
@@ -50,219 +133,270 @@ def my_cylinder2():
...
+            TempFluidDynamics(
+                entities=[
+                    my_cylinder1,
+                    my_cylinder1,
+                    my_cylinder1,
+                    my_volume_mesh1["*"],
+                    my_volume_mesh1["*zone*"],
+                ]
+            ),
+            TempWallBC(surfaces=[my_volume_mesh1["*"], "*", "farfield"]),
```
Comment (benflexcompute) [inline tests/simulation_framework/test_entities_v2.py:382]: This "*" actually raises concern that the user might use "*" thinking it is going to match everything when in reality it is only going to match surfaces in the Param's registry that contains implicitly defined surfaces.

File: tests/simulation_framework/test_entities_v2.py [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+        self.fname = file_name
+        self._populate_registry()
+
+
+class TempSurface(_SurfaceEntityBase):
+    pass
+
+
+class TempFluidDynamics(Flow360BaseModel):
+    entities: EntityList[GenericVolume, Box, Cylinder, str] = pd.Field(alias="volumes", default=[])
```
[the remark this answers is not in the document]
Comment (maciej-flexcompute) [reply tests/simulation_framework/test_entities_v2.py:76]: Maybe we can call it EntityUnion?

File: flow360/component/simulation/framework/entity_base.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+    Attributes:
+        _entity_type (str): A string representing the specific type of the entity.
+                            This should be set in subclasses to differentiate between entity types.
+                            Note this controls the granularity of the registry.
+        _is_generic(bool): A flag indicating whether the entity is a generic entity (constructed from metadata).
+        name (str): The name of the entity, used for identification and retrieval.
+    """
+
+    _entity_type: str = None
```
[the remark this answers is not in the document]
Comment (benflexcompute) [reply flow360/component/simulation/framework/entity_base.py:27]: I think names with leading underscore automatically converts to PrivateAttr: 
https://docs.pydantic.dev/dev/concepts/models/#private-model-attributes:~:text=Attributes%20whose%20name%20has%20a%20leading%20underscore%20are%20not%20treated%20as%20fields%20by%20Pydantic%2C%20and%20are%20not%20included%20in%20the%20model%20schema.%20Instead%2C%20these%20are%20converted%20into%20a%20%22private%20attribute%22%20which%20is%20not%20validated%20or%20even%20set%20during%20calls%20to%20__init__%2C%20model_validate%2C%20etc.

They have an example here: https://docs.pydantic.dev/dev/concepts/models/#private-model-attributes:~:text=.now)-,_secret_value%3A%20str,-def%20__init__(

File: flow360/component/simulation/framework/entity_base.py [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+    _is_generic = False
```
[the remark this answers is not in the document]
Comment (benflexcompute) [reply flow360/component/simulation/framework/entity_base.py:28]: Renamed to `_auto_constructed` and added more docstring

File: flow360/component/simulation/framework/entity_base.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+
+    def __init__(self, **data):
+        """
+        Initializes a new entity and registers it in the global registry.
+
+        Parameters:
+            data: Keyword arguments containing initial values for fields declared in the entity.
+        """
+        super().__init__(**data)
+        assert self._entity_type is not None, "_entity_type is not defined in the entity class."
```
[the remark this answers is not in the document]
Comment (benflexcompute) [reply flow360/component/simulation/framework/entity_base.py:39]: It is [not possible](https://docs.pydantic.dev/dev/api/fields/#pydantic.fields.PrivateAttr:~:text=normal%20validation/serialization.-,Private%20attributes%20are%20not%20validated%20by%20Pydantic%2C%20so%20it%27s%20up%20to%20you%20to%20ensure%20they%20are%20used%20in%20a%20type%2Dsafe%20manner.,-Private%20attributes%20are) to specify field validator for private attribute. They are basically ignored by Pydantic. I changed this to throw NotImplementedError.

File: flow360/component/simulation/framework/entity_base.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+            )
+        if "name" not in update or update["name"] == self.name:
+            raise ValueError(
+                "Copying an entity requires a new name to be specified. Please provide a new name in the update dictionary."
+            )
+        return super().copy(update=update, **kwargs)
+
+
+class _CombinedMeta(type(Flow360BaseModel), type):
+    pass
```
[the remark this answers is not in the document]
Comment (benflexcompute) [reply flow360/component/simulation/framework/entity_base.py:64]: If it only inherit from type then python gives error: `TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases`

File: flow360/component/simulation/framework/entity_base.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+
+
+class _EntitiesListMeta(_CombinedMeta):
```
[the remark this answers is not in the document]
Comment (benflexcompute) [reply flow360/component/simulation/framework/entity_base.py:67]: Done

File: flow360/component/simulation/framework/entity_base.py [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+            and hasattr(entity_field_type, "__origin__")
+            and entity_field_type.__origin__ is list
+        ):
+            valid_types = get_args(entity_field_type)[0]
+            if hasattr(valid_types, "__origin__") and valid_types.__origin__ is Union:
+                valid_types = get_args(valid_types)
+            else:
+                valid_types = (valid_types,)
+            return valid_types
+        raise TypeError("Internal error, the metaclass for EntityList is not properly set.")
```
[the remark this answers is not in the document]
Comment (benflexcompute) [reply flow360/component/simulation/framework/entity_base.py:144]: Feature --> entities
SurfaceOutput --> surfaces
SliceOutput --> slices
IsoSurface --> isoSurface
SurfaceIntegralMonitor --> surfaceIntegralMonitor
ProbeMonitor --> probeMonitor
Rotation --> Union[GenericZone, Cylinder]
FluidDynamics --> Uniton[**all volume types**]
BETDisk --> Union[**same as Rotation**]
......
The gist is that there are many possibilities. This is why I choose it to be dynamic.

File: flow360/component/simulation/framework/entity_base.py [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+        """
+        if not isinstance(entity_types, tuple):
+            entity_types = (entity_types,)
+        union_type = Union[entity_types]
+        annotations = {"stored_entities": List[union_type]}
+        new_cls = type(
+            f"{cls.__name__}[{','.join([t.__name__ for t in entity_types])}]",
+            (cls,),
+            {"__annotations__": annotations},
+        )
```
[the remark this answers is not in the document]
Comment (benflexcompute) [reply flow360/component/simulation/framework/entity_base.py:80]: Yes it is. This is to support syntax like: `EntityList[GenericVolume, Box, Cylinder, str]` so developers can easily specify what are allowed for that particular "entities" field.

File: flow360/component/simulation/primitives.py [hunk 11]
```diff
@@ -14,15 +14,51 @@ class Transformation(Flow360BaseModel):
...
+
+
+class GenericSurface(_SurfaceEntityBase):
+    """Do not expose"""
+
+    _is_generic = True
+
+    def __init_subclass__(cls, *args, **kwargs):
+        super().__init_subclass__(*args, **kwargs)
+        raise TypeError(f"Subclassing of {cls.__name__} is not allowed")
```
[the remark this answers is not in the document]
Comment (benflexcompute) [reply flow360/component/simulation/primitives.py:42]: The main reason is trying to stop developers **_accidentally_** change the `_entity_type` (which affects the registry bucketing) or `_is_generic` (which affects whether they will be removed if duplicate is found). 

I know even if I stop subclassing them I cannot stop developer from changing these classes themselves. I added some description on it.

File: flow360/component/simulation/primitives.py [hunk 12]
```diff
@@ -14,15 +14,51 @@ class Transformation(Flow360BaseModel):
...
+
+
+class Box(_VolumeEntityBase):
+    center: Tuple[float, float, float] = pd.Field()
     size: Tuple[float, float, float] = pd.Field()
     axes: Tuple[Tuple[float, float, float], Tuple[float, float, float]] = pd.Field()
 
+    def __init_subclass__(cls, *args, **kwargs):
+        super().__init_subclass__(*args, **kwargs)
+        raise TypeError(f"Subclassing of {cls.__name__} is not allowed")
```
[the remark this answers is not in the document]
Comment (benflexcompute) [reply flow360/component/simulation/primitives.py:52]: See above.

File: tests/simulation_framework/test_entities_v2.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (benflexcompute) [reply tests/simulation_framework/test_entities_v2.py:76]: I see your point but it is a list by type. The union part is a bit self-explanatory looking at the []  syntax. Or we call it EntityUnionList?

One thing to point out is that entity list is[ usually associated with](https://en.wikipedia.org/wiki/Entity_List) the trade restriction list published by the United States Department of Commerce so maybe it is a bit sensitive to call it EntityList?

File: flow360/component/simulation/framework/entity_base.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (maciej-flexcompute) [reply flow360/component/simulation/framework/entity_base.py:31]: good to know!

**Review by maciej-flexcompute — APPROVED** [review]

**Pull request merged**